File: Controllers/TodoController.cs
using Microsoft.AspNetCore.Mvc;
using TodoDemo.Models;
using TodoDemo.Services;

namespace TodoDemo.Controllers;

public class TodoController : Controller
{
    private readonly TodoService _todoService;

    public TodoController(TodoService todoService)
    {
        _todoService = todoService;
    }

    // GET: Todo
    public IActionResult Index()
    {
        var todos = _todoService.GetAll();
        return View(todos);
    }

    // GET: Todo/Create
    public IActionResult Create()
    {
        return View();
    }

    // POST: Todo/Create
    [HttpPost]
    [ValidateAntiForgeryToken]
    public IActionResult Create(Todo todo)
    {
        if (ModelState.IsValid)
        {
            _todoService.Add(todo);
            return RedirectToAction(nameof(Index));
        }
        return View(todo);
    }

    // GET: Todo/Edit/5
    public IActionResult Edit(int id)
    {
        var todo = _todoService.GetById(id);
        if (todo == null)
        {
            return NotFound();
        }
        return View(todo);
    }

    // POST: Todo/Edit/5
    [HttpPost]
    [ValidateAntiForgeryToken]
    public IActionResult Edit(int id, Todo todo)
    {
        if (id != todo.Id)
        {
            return NotFound();
        }

        if (ModelState.IsValid)
        {
            _todoService.Update(todo);
            return RedirectToAction(nameof(Index));
        }
        return View(todo);
    }

    // POST: Todo/Delete/5
    [HttpPost]
    [ValidateAntiForgeryToken]
    public IActionResult Delete(int id)
    {
        _todoService.Delete(id);
        return RedirectToAction(nameof(Index));
    }

    // POST: Todo/Toggle/5
    [HttpPost]
    [ValidateAntiForgeryToken]
    public IActionResult Toggle(int id)
    {
        _todoService.ToggleComplete(id);
        return RedirectToAction(nameof(Index));
    }
}

File: java-todo-app/src/main/java/com/tododemo/controller/TodoController.java
package com.tododemo.controller;

import com.tododemo.model.Todo;
import com.tododemo.service.TodoService;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.validation.BindingResult;
import org.springframework.web.bind.annotation.*;

import java.util.Optional;

@Controller
@RequestMapping("/")
public class TodoController {

    private final TodoService todoService;

    @Autowired
    public TodoController(TodoService todoService) {
        this.todoService = todoService;
    }

    /**
     * Display all todos
     */
    @GetMapping
    public String index(Model model) {
        var todos = todoService.getAll();
        model.addAttribute("todos", todos);
        model.addAttribute("totalCount", todos.size());
        model.addAttribute("completedCount", todos.stream().filter(Todo::isCompleted).count());
        model.addAttribute("pendingCount", todos.stream().filter(t -> !t.isCompleted()).count());
        return "index";
    }

    /**
     * Show create form
     */
    @GetMapping("/create")
    public String create(Model model) {
        model.addAttribute("todo", new Todo());
        return "create";
    }

    /**
     * Handle create form submission
     */
    @PostMapping("/create")
    public String createPost(@Valid @ModelAttribute("todo") Todo todo, 
                            BindingResult bindingResult) {
        if (bindingResult.hasErrors()) {
            return "create";
        }
        todoService.add(todo);
        return "redirect:/";
    }

    /**
     * Show edit form
     */
    @GetMapping("/edit/{id}")
    public String edit(@PathVariable int id, Model model) {
        Optional<Todo> todo = todoService.getById(id);
        if (todo.isEmpty()) {
            return "redirect:/";
        }
        model.addAttribute("todo", todo.get());
        return "edit";
    }

    /**
     * Handle edit form submission
     */
    @PostMapping("/edit/{id}")
    public String editPost(@PathVariable int id, 
                          @Valid @ModelAttribute("todo") Todo todo,
                          BindingResult bindingResult) {
        if (bindingResult.hasErrors()) {
            return "edit";
        }
        if (id != todo.getId()) {
            return "redirect:/";
        }
        todoService.update(todo);
        return "redirect:/";
    }

    /**
     * Handle delete action
     */
    @PostMapping("/delete/{id}")
    public String delete(@PathVariable int id) {
        todoService.delete(id);
        return "redirect:/";
    }

    /**
     * Handle toggle completion action
     */
    @PostMapping("/toggle/{id}")
    public String toggle(@PathVariable int id) {
        todoService.toggleComplete(id);
        return "redirect:/";
    }
}
